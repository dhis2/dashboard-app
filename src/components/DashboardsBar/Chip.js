import { useDhis2ConnectionStatus } from '@dhis2/app-runtime'
import { Chip as UiChip, colors, IconStarFilled24 } from '@dhis2/ui'
import cx from 'classnames'
import debounce from 'lodash/debounce.js'
import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router-dom'
import { apiPostDataStatistics } from '../../api/dataStatistics.js'
import { useCacheableSection } from '../../modules/useCacheableSection.js'
import { OfflineSaved } from './assets/icons.js'
import classes from './styles/Chip.module.css'

const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const { lastUpdated } = useCacheableSection(dashboardId)
    const { isConnected: online } = useDhis2ConnectionStatus()
    const chipProps = {
        selected,
    }

    if (starred) {
        chipProps.icon = (
            <IconStarFilled24
                color={selected ? colors.white : colors.grey600}
            />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        online && debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>
                <span
                    className={
                        lastUpdated ? classes.labelWithAdornment : undefined
                    }
                >
                    {label}
                </span>
                {lastUpdated && (
                    <OfflineSaved
                        className={cx(
                            classes.adornment,
                            selected && classes.selected
                        )}
                    />
                )}
            </UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default Chip
