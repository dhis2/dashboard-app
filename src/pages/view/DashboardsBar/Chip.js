import React from 'react'
import PropTypes from 'prop-types'
import { Chip as UiChip, colors, IconStarFilled24 } from '@dhis2/ui'
import { Link } from 'react-router-dom'
import debounce from 'lodash/debounce'

import { apiPostDataStatistics } from '../../../api/dataStatistics'

import classes from './styles/Chip.module.css'

export const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const chipProps = {
        selected,
    }

    if (starred) {
        chipProps.icon = (
            <IconStarFilled24
                color={selected ? colors.white : colors.grey600}
            />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>{label}</UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    classes: PropTypes.object,
}

export default Chip
