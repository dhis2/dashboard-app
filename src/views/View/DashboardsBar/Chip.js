import React from 'react'
import PropTypes from 'prop-types'
import { Chip as UiChip } from '@dhis2/ui'
import { Link } from 'react-router-dom'
import debounce from 'lodash/debounce'

import StarIcon from './assets/Star'
import { apiPostDataStatistics } from '../../../api/dataStatistics'

import classes from './styles/Chip.module.css'

export const Chip = ({ starred, selected, label, dashboardId, onClick }) => {
    const chipProps = {
        selected,
    }

    if (starred) {
        const selectedState = selected ? classes.selected : classes.unselected
        chipProps.icon = (
            <StarIcon className={`${classes.icon} ${selectedState}`} />
        )
    }
    const debouncedPostStatistics = debounce(
        () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
        500
    )

    const handleClick = () => {
        debouncedPostStatistics()
        onClick()
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={handleClick}
            data-test="dashboard-chip"
        >
            <UiChip {...chipProps}>{label}</UiChip>
        </Link>
    )
}

Chip.propTypes = {
    dashboardId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool.isRequired,
    starred: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    classes: PropTypes.object,
}

export default Chip
