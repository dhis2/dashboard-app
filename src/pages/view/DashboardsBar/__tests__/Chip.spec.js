import React from 'react'
import { render } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import Chip from '../Chip'

const defaultProps = {
    onClick: jest.fn(),
    label: 'Hello Rainbow Dash',
    dashboardId: 'myLittlePony',
}

test('Chip renders unstarred and unselected', () => {
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} starred={false} selected={false} />
        </Router>
    )
    expect(container).toMatchSnapshot()
})

test('Chip renders starred and unselected', () => {
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} starred={true} selected={false} />
        </Router>
    )
    expect(container).toMatchSnapshot()
})

test('Chip renders starred and selected', () => {
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} starred={true} selected={true} />
        </Router>
    )
    expect(container).toMatchSnapshot()
})

test('Chip renders unstarred and selected', () => {
    const { container } = render(
        <Router history={createMemoryHistory()}>
            <Chip {...defaultProps} starred={false} selected={true} />
        </Router>
    )
    expect(container).toMatchSnapshot()
})
