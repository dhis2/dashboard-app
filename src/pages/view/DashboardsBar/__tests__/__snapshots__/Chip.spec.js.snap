// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip renders starred and selected 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/myLittlePony"
  >
    <span
      class="jsx-3285323350 selected"
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#ffffff"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        Hello Rainbow Dash
      </span>
    </span>
  </a>
</div>
`;

exports[`Chip renders starred and unselected 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/myLittlePony"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-3489677657"
        data-test="dhis2-uicore-chip-icon"
      >
        <svg
          color="#6e7a8a"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.397 7.041l-6.051.88-.114.022a1 1 0 00-.44 1.683l4.378 4.268-1.033 6.027-.013.11a1 1 0 001.463.944L12 18.13l5.413 2.845.1.047a1 1 0 001.35-1.1l-1.034-6.028 4.38-4.268.078-.086a1 1 0 00-.633-1.62l-6.052-.879-2.705-5.484a1 1 0 00-1.794 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="jsx-785997980 "
      >
        Hello Rainbow Dash
      </span>
    </span>
  </a>
</div>
`;

exports[`Chip renders unstarred and selected 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/myLittlePony"
  >
    <span
      class="jsx-3285323350 selected"
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-785997980 "
      >
        Hello Rainbow Dash
      </span>
    </span>
  </a>
</div>
`;

exports[`Chip renders unstarred and unselected 1`] = `
<div>
  <a
    class="link"
    data-test="dashboard-chip"
    href="/myLittlePony"
  >
    <span
      class="jsx-3285323350 "
      data-test="dhis2-uicore-chip"
    >
      <span
        class="jsx-785997980 "
      >
        Hello Rainbow Dash
      </span>
    </span>
  </a>
</div>
`;
